Source/ACE/Entity/Player.cs

```diff
@@ -236,20 +236,77 @@ public Player(Session session) : base(ObjectType.Creature, session.CharacterRequ
             Session.WorldSession.EnqueueSend(setTurbineChatChannels, general, trade, lfg, roleplay);
         }
 
+        /// <summary>
+        /// Raise the available XP by a specified amount
+        /// </summary>
+        /// <param name="amount">A unsigned long containing the desired XP amount to raise</param>
         public void GrantXp(ulong amount)
         {
+            var test = amount.ToString().Length;
             character.GrantXp(amount);
+            CheckForLevelup();
             var xpTotalUpdate = new GameMessagePrivateUpdatePropertyInt64(Session, PropertyInt64.TotalExperience, character.TotalExperience);
             var xpAvailUpdate = new GameMessagePrivateUpdatePropertyInt64(Session, PropertyInt64.AvailableExperience, character.AvailableExperience);
             var message = new GameMessageSystemChat($"{amount} experience granted.", ChatMessageType.Broadcast);
             Session.WorldSession.EnqueueSend(xpTotalUpdate, xpAvailUpdate, message);
         }
 
+        /// <summary>
+        /// Determines if the player has advanced a level
+        /// </summary>
+        /// <remarks>
+        /// Known issues:
+        ///         1. XP updates from outside of the grantxp command have not been done yet.
+        /// </remarks>
         private void CheckForLevelup()
         {
+            // Question: Where do *we* call CheckForLevelup()? : 
+            //      From within the player.cs file, the options might be:
+            //           GrantXp()

```

Comment:
Makes sense to me as I see it now... I think this function will expand a bit further to refine where the Xp came from, and where to apply it to.. Thinking for things like xp passup as well as xp coming from using skills and even the grantxp command..

list[0] = LF_ENUMERATE, public, value = 0, name = 'Undef_ExperienceHandlingType'
	list[1] = LF_ENUMERATE, public, value = 1, name = 'ApplyLevelMod_ExperienceHandlingType'
	list[2] = LF_ENUMERATE, public, value = 2, name = 'ShareWithFellows_ExperienceHandlingType'
	list[3] = LF_ENUMERATE, public, value = 4, name = 'AddFellowshipBonus_ExperienceHandlingType'
	list[4] = LF_ENUMERATE, public, value = 8, name = 'ShareWithAllegiance_ExperienceHandlingType'
	list[5] = LF_ENUMERATE, public, value = 16, name = 'ApplyToVitae_ExperienceHandlingType'
	list[6] = LF_ENUMERATE, public, value = 32, name = 'EarnsCP_ExperienceHandlingType'
	list[7] = LF_ENUMERATE, public, value = 64, name = 'ReducedByDistance_ExperienceHandlingType'
	list[8] = LF_ENUMERATE, public, value = 95, name = 'Monster_ExperienceHandlingType'
	list[9] = LF_ENUMERATE, public, value = 26, name = 'NormalQuest_ExperienceHandlingType'
	list[10] = LF_ENUMERATE, public, value = 16, name = 'NoShareQuest_ExperienceHandlingType'
	list[11] = LF_ENUMERATE, public, value = 24, name = 'PassupQuest_ExperienceHandlingType'
	list[12] = LF_ENUMERATE, public, value = 24, name = 'ReceivedFromFellowship_ExperienceHandlingType'
	list[13] = LF_ENUMERATE, public, value = 127, name = 'PPEarnedFromUse_ExperienceHandlingType'
	list[14] = LF_ENUMERATE, public, value = 16, name = 'AdminRaiseXP_ExperienceHandlingType'
	list[15] = LF_ENUMERATE, public, value = 16, name = 'AdminRaiseSkillXP_ExperienceHandlingType'
	list[16] = LF_ENUMERATE, public, value = 0, name = 'ReceivedFromAllegiance_ExperienceHandlingType'
	list[17] = LF_ENUMERATE, public, value = (LF_ULONG) 2147483647, name = 'FORCE_ExperienceHandlingType_32_BIT'